Apply C++ type safety to Lookup.h's HashTableValue's ValueStorage union.

MenloDorian opened the pull request:
#### 757f991622c717f67e075f197f7116b431a5376a
<pre>
Apply C++ type safety to Lookup.h&apos;s HashTableValue&apos;s ValueStorage union.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243680">https://bugs.webkit.org/show_bug.cgi?id=243680</a>
&lt;rdar://problem/98206776&gt;

Reviewed by Saam Barati.

1. Instead of always casting all the potential union field values to/from intptr_t values,
   we now strongly type each of the union variants.

   We also provide constructors for each of the variants using enum tags to distinguish
   between which variant we want to initialize the union with.  create_hash_table and
   CodeGeneratorJS.pm has been updated to emit these enum tags as well eliding the now
   unneeded casts for each of the initialization arguments.

   Also added a Debug ASSERT to double check that we aren&apos;t consuming the union as one
   variant if it was initialized as another variant.

2. Factored FunctionPtr out of MacroAssemblerCodeRef.h into the WTF::TypedFunctionPtr class.
   TypedFunctionPtr is type safe with regards to the function prototype i.e. it preserved
   type info about the return value and arguments.

   TypedFunctionPtr can be used to call the function that it points to.

   We now use this for PropertySlot::GetValueFunc and PropertySlot::PutValueFunc for
   some added PAC diversity, and also as a debugging aid for the HashTableValue union
   variants.

3. Also introduced a WTF::FunctionPtr, which erases the function prototype info
   by always casting the stored pointer to a void(*)(int) pointer.

   FunctionPtr cannot be used to call the function that it points to (because its type
   Info is erased).

   In order to use it in a call, the FunctionPtr needs to be converted back into a
   normal function pointer (os a TypedFunctionPtr) first.

4. NativeFunction is now implemented using TypedFunctionPtr.

   RawNativeFunction has been removed.  Its role is subsumed by NativeFunction::Ptr.
   The only exception to this is one place in the JSC public API where it defines its own
   RawNativeFunction in its private implementation.  This was not changed to just so we
   don&apos;t have to #include wtf/FunctionPtr.h in that public API file.

   TaggedNativeFunction is now implemented using the type erased FunctionPtr.

5. CFunctionPtr (in MacroAssemblerCodeRef.h) has been removed.  It was only used by
   DOMJIT::Signature.  Instead, DOMJIT::Signature now defines its own FunctionPtr
   using WTF::FunctionPtr.

6. Removed all the aliases of GetValueFunc and PutValueFunc in PropertySlot, PutPropertySlot,
   etc.  instead, just have all the places use GetValueFunc and PutValueFunc directly.
   There&apos;s no value added by having all these aliases.

* Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
(JSC::JSCallbackObject&lt;JSAPIWrapperObject&gt;::getCallFunction):
(JSC::JSCallbackObject&lt;JSAPIWrapperObject&gt;::getConstructFunction):
* Source/JavaScriptCore/API/JSCallbackObject.cpp:
(JSC::JSCallbackObject&lt;JSNonFinalObject&gt;::getCallFunction):
(JSC::JSCallbackObject&lt;JSNonFinalObject&gt;::getConstructFunction):
(JSC::JSCallbackObject&lt;JSGlobalObject&gt;::getCallFunction):
(JSC::JSCallbackObject&lt;JSGlobalObject&gt;::getConstructFunction):
* Source/JavaScriptCore/API/glib/JSAPIWrapperGlobalObject.cpp:
(JSC::JSCallbackObject&lt;JSAPIWrapperGlobalObject&gt;::getCallFunction):
(JSC::JSCallbackObject&lt;JSAPIWrapperGlobalObject&gt;::getConstructFunction):
* Source/JavaScriptCore/API/glib/JSAPIWrapperObjectGLib.cpp:
(JSC::JSCallbackObject&lt;JSAPIWrapperObject&gt;::getCallFunction):
(JSC::JSCallbackObject&lt;JSAPIWrapperObject&gt;::getConstructFunction):
* Source/JavaScriptCore/assembler/LinkBuffer.h:
(JSC::LinkBuffer::link):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::readCallTarget):
(JSC::MacroAssemblerARM64::linkCall):
* Source/JavaScriptCore/assembler/MacroAssemblerCodeRef.h:
(JSC::toFunctionPtr):
(JSC::CFunctionPtr::CFunctionPtr): Deleted.
(JSC::CFunctionPtr::get const): Deleted.
(JSC::CFunctionPtr::address const): Deleted.
(JSC::CFunctionPtr::operator bool const): Deleted.
(JSC::CFunctionPtr::operator! const): Deleted.
(JSC::CFunctionPtr::operator== const): Deleted.
(JSC::CFunctionPtr::operator!= const): Deleted.
(JSC::FunctionPtr::FunctionPtr): Deleted.
(JSC::FunctionPtr::retagged const): Deleted.
(JSC::FunctionPtr::executableAddress const): Deleted.
(JSC::FunctionPtr::retaggedExecutableAddress const): Deleted.
(JSC::FunctionPtr::operator bool const): Deleted.
(JSC::FunctionPtr::operator! const): Deleted.
(JSC::FunctionPtr::operator== const): Deleted.
(JSC::FunctionPtr::operator!= const): Deleted.
(JSC::FunctionPtr&lt;tag&gt;::FunctionPtr): Deleted.
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::ftlThunkAwareRepatchCall):
(JSC::linkPolymorphicCall):
* Source/JavaScriptCore/create_hash_table:
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::compileFunction):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/domjit/DOMJITGetterSetter.h:
* Source/JavaScriptCore/domjit/DOMJITSignature.h:
(JSC::DOMJIT::Signature::Signature):
(JSC::DOMJIT::Signature::functionWithoutTypeCheck const):
* Source/JavaScriptCore/ftl/FTLLink.cpp:
(JSC::FTL::link):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::emitNakedNearCall):
(JSC::JIT::emitNakedNearTailCall):
* Source/JavaScriptCore/runtime/ExecutableBase.cpp:
(JSC::ExecutableBase::dump const):
* Source/JavaScriptCore/runtime/InternalFunction.cpp:
(JSC::InternalFunction::InternalFunction):
* Source/JavaScriptCore/runtime/JSCPtrTag.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.h:
* Source/JavaScriptCore/runtime/Lookup.h:
(JSC::HashTableValue::builtinGenerator const):
(JSC::HashTableValue::function const):
(JSC::HashTableValue::functionLength const):
(JSC::HashTableValue::propertyGetter const):
(JSC::HashTableValue::propertyPutter const):
(JSC::HashTableValue::domJIT const):
(JSC::HashTableValue::signature const):
(JSC::HashTableValue::accessorGetter const):
(JSC::HashTableValue::accessorSetter const):
(JSC::HashTableValue::constantInteger const):
(JSC::HashTableValue::lexerValue const):
(JSC::HashTableValue::lazyCellPropertyOffset const):
(JSC::HashTableValue::lazyClassStructureOffset const):
(JSC::HashTableValue::lazyPropertyCallback const):
(JSC::HashTableValue::builtinAccessorGetterGenerator const):
(JSC::HashTableValue::builtinAccessorSetterGenerator const):
* Source/JavaScriptCore/runtime/NativeFunction.h:
(JSC::toTagged):
(WTF::add):
(JSC::NativeFunction::NativeFunction): Deleted.
(JSC::NativeFunction::operator intptr_t const): Deleted.
(JSC::NativeFunction::operator bool const): Deleted.
(JSC::NativeFunction::operator! const): Deleted.
(JSC::NativeFunction::operator== const): Deleted.
(JSC::NativeFunction::operator!= const): Deleted.
(JSC::NativeFunction::operator()): Deleted.
(JSC::NativeFunction::rawPointer const): Deleted.
(JSC::TaggedNativeFunction::TaggedNativeFunction): Deleted.
(JSC::TaggedNativeFunction::operator bool const): Deleted.
(JSC::TaggedNativeFunction::operator! const): Deleted.
(JSC::TaggedNativeFunction::operator== const): Deleted.
(JSC::TaggedNativeFunction::operator!= const): Deleted.
(JSC::TaggedNativeFunction::operator()): Deleted.
(JSC::TaggedNativeFunction::operator NativeFunction): Deleted.
(JSC::TaggedNativeFunction::rawPointer const): Deleted.
* Source/JavaScriptCore/runtime/PropertySlot.h:
(JSC::PropertySlot::setCacheableCustom):
* Source/JavaScriptCore/runtime/PutPropertySlot.h:
(JSC::PutPropertySlot::setCustomValue):
(JSC::PutPropertySlot::setCustomAccessor):
(JSC::PutPropertySlot::customSetter const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::getHostFunction):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FunctionPtr.h: Added.
(WTF::TypedFunctionPtr::TypedFunctionPtr):
(WTF::TypedFunctionPtr::operator() const):
(WTF::TypedFunctionPtr::get const):
(WTF::TypedFunctionPtr:: const):
(WTF::TypedFunctionPtr::executableAddress const):
(WTF::TypedFunctionPtr::retaggedExecutableAddress const):
(WTF::TypedFunctionPtr::untaggedExecutableAddress const):
(WTF::TypedFunctionPtr::operator bool const):
(WTF::TypedFunctionPtr::operator! const):
(WTF::TypedFunctionPtr::operator== const):
(WTF::TypedFunctionPtr::operator!= const):
(WTF::TypedFunctionPtr::operator=):
(WTF::TypedFunctionPtr::encode):
(WTF::TypedFunctionPtr::decode):
(WTF::FunctionPtr::FunctionPtr):
(WTF::FunctionPtr::retagged const):
(WTF::add):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::createCrossOriginFunction):
(WebCore::JSDOMGlobalObject::createCrossOriginGetterSetter):
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::jsDOMWindowGetOwnPropertySlotRestrictedAccess):
(WebCore::JSDOMWindow::put):
* Source/WebCore/bindings/js/JSLocationCustom.cpp:
(WebCore::JSLocation::put):
* Source/WebCore/bindings/js/JSRemoteDOMWindowCustom.cpp:
(WebCore::JSRemoteDOMWindow::put):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHashTableValueArray):
* Source/WebCore/bindings/scripts/test/JS/JSDOMWindow.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSDedicatedWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSExposedStar.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSExposedToWorkerAndWindow.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSPaintWorkletGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSServiceWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSShadowRealmGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSSharedWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCEReactions.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCEReactionsStringifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCallTracer.cpp:
(WebCore::JSTestCallTracerDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestClassWithJSBuiltinConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestConditionalIncludes.cpp:
(WebCore::JSTestConditionalIncludesDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditionallyReadWrite.cpp:
(WebCore::JSTestConditionallyReadWriteDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestDOMJIT.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSON.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONFilteredByExposed.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONIndirectInheritance.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInherit.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInheritFinal.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDelegateToSharedSyntheticAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDomainSecurity.cpp:
(WebCore::JSTestDomainSecurityDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestEnabledBySetting.cpp:
(WebCore::JSTestEnabledBySettingDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestEnabledForContext.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestEventConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestEventTarget.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGenerateAddOpaqueRoot.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGenerateIsReachable.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestInterface.cpp:
(WebCore::JSTestInterfaceDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSTestInterfaceLeadingUnderscore.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIterable.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestJSBuiltinConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyFactoryFunction.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyNoInterfaceObject.cpp:
(WebCore::JSTestLegacyNoInterfaceObjectPrototype::finishCreation):
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyOverrideBuiltIns.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestMapLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestMapLikeWithOverriddenOperations.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterWithIndexedGetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterCallWith.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIndexedGetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIndexedGetterAndSetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyOverrideBuiltIns.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyUnforgeableProperties.cpp:
(WebCore::JSTestNamedSetterWithLegacyUnforgeablePropertiesDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyUnforgeablePropertiesAndLegacyOverrideBuiltIns.cpp:
(WebCore::JSTestNamedSetterWithLegacyUnforgeablePropertiesAndLegacyOverrideBuiltInsDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestNamespaceConst.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamespaceObject.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSTestOperationConditional.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestOverloadedConstructors.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestOverloadedConstructorsWithSequence.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestPluginInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestPromiseRejectionEvent.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReadOnlyMapLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReadOnlySetLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReportExtraMemoryCost.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSerializedScriptValueInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSetLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSetLikeWithOverriddenOperations.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierAnonymousOperation.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierNamedOperation.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierOperationImplementedAs.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierOperationNamedToString.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierReadOnlyAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierReadWriteAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestTaggedWrapper.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestTypedefs.cpp:
(WebCore::JSTestTypedefsDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSWorkletGlobalScope.cpp:
* Source/WebCore/cssjit/FunctionCall.h:
(WebCore::FunctionCall::FunctionCall):
(WebCore::FunctionCall::setFunctionAddress):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementAttributeFunctionCallValueMatching):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementFunctionCallTest):

Canonical link: <a href="https://commits.webkit.org/253414@main">https://commits.webkit.org/253414@main</a>
</pre>

saambarati commented on Source/JavaScriptCore/runtime/JSCPtrTag.h:
```diff
@@ -52,6 +51,10 @@ using PtrTag = WTF::PtrTag;
     /* Callee:Native Caller:Native */ \
     v(BytecodePtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
     v(CustomAccessorPtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
+    v(GetValueFuncPtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
+    v(GetValueFuncWithPtrPtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
+    v(PutValueFuncPtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
+    v(PutValueFuncWithPtrPtrTag, PtrTagCalleeType::Native, PtrTagCallerType::Native) \
```
Don't we call these from JIT code too?

MenloDorian replied:
I don't think so.  Anything called by the JIT needs to be tagged with OperationPtrTag.  Previously, these were always tagged with CFunctionPtrTag, which should not be called directly from the JIT.

saambarati commented on Source/WTF/wtf/FunctionPtr.h:
```diff
@@ -0,0 +1,228 @@
+    { }
+#endif
+
+    Out operator()(In... in) const
```
We should use perfect forwarding here. I believe we want:
```
Out operator()(In&&... in) const
```

MenloDorian replied:
`Out operator()(In&&... in) const` doesn't work.  class `Function` also does not use it.  I still haven't figured out why yet.  The error message implies that it's expecting to forward just one argument rather than all the arguments.  Will look more later.

saambarati commented on Source/WTF/wtf/FunctionPtr.h:
```diff
@@ -0,0 +1,228 @@
+#endif
+
+template<PtrTag tag, typename Out, typename... In>
+class TypedFunctionPtr {
```
Should we call this TypedCFunctionPtr? We're expecting the incoming pointer to be tagged with a C function pointer

saambarati replied:
Also, I think some implementation of this class becomes easier if we just sign null

saambarati replied:
My suggestion to sign null might be worse for perf, though. Maybe it's not worth it

MenloDorian replied:
I'm going to keep the null checks.

saambarati commented on Source/WTF/wtf/FunctionPtr.h:
```diff
@@ -0,0 +1,228 @@
+            return ptr;
+        if (!ptr)
+            return nullptr;
+        assertIsNullOrCFunctionPtr(ptr);
```
we should assert this at the top of the function

MenloDorian replied:
I cannot assert this at the top because it breaks usage of this function as `constexpr`.  I'll change it to `assertIsCFunctionPtr` there.

saambarati commented on Source/JavaScriptCore/runtime/PropertySlot.h:
```diff
@@ -85,11 +86,11 @@ inline unsigned attributesForStructure(unsigned attributes)
 
-using PutValueFunc = bool (JIT_OPERATION_ATTRIBUTES*)(JSGlobalObject*, EncodedJSValue baseObject, EncodedJSValue value, PropertyName);
-using PutValueFuncWithPtr = bool (JIT_OPERATION_ATTRIBUTES*)(JSGlobalObject*, EncodedJSValue baseObject, EncodedJSValue value, PropertyName, void*);
+using PutValueFunc = TypedFunctionPtr<PutValueFuncPtrTag, bool, JSGlobalObject*, EncodedJSValue, EncodedJSValue, PropertyName>;
```
From what I said in slack, but want to put it into code review, too. Would be nice if we templatized `TypedFunctionPtr` like we do `Function`, so this LOC can become more readable by writing:
`TypedFunctionPtr<PutValueFuncPtrTag, bool(JSGlobalObject*, EncodedJSValue, EncodedJSValue, PropertyName)>`

MenloDorian replied:
Done.

saambarati commented on Source/JavaScriptCore/runtime/NativeFunction.h:
```diff
@@ -105,25 +53,29 @@ struct TaggedNativeFunctionHash {
 inline void add(Hasher& hasher, JSC::NativeFunction key)
 {
-    add(hasher, bitwise_cast<uintptr_t>(key));
+    add(hasher, key.executableAddress());
```
We can't hash the signed value?

MenloDorian replied:
`executableAddress()` is the signed value.

saambarati commented on Source/WTF/wtf/FunctionPtr.h:
```diff
@@ -0,0 +1,228 @@
+template<PtrTag tag>
+class FunctionPtr : public TypedFunctionPtr<tag, void, int> {
+    using Base = TypedFunctionPtr<tag, void, int>;
+    using PtrType = void(*)(int);
```
Why treat it as a function pointer at all, instead of just void*?

MenloDorian replied:
Because the `TypedFunctionPtr` only handles function pointers.  Since `FunctionPtr` here is re-using `TypedFunctionPtr`'s implementation, it needs to play by those rules ... not that `FunctionPtr` really cares how the underlying bits are represented.

saambarati reviewed (approved):
r=me with comments

Commit:
Apply C++ type safety to Lookup.h's HashTableValue's ValueStorage union.
https://bugs.webkit.org/show_bug.cgi?id=243680
<rdar://problem/98206776>

Reviewed by Saam Barati.

1. Instead of always casting all the potential union field values to/from intptr_t values,
   we now strongly type each of the union variants.

   We also provide constructors for each of the variants using enum tags to distinguish
   between which variant we want to initialize the union with.  create_hash_table and
   CodeGeneratorJS.pm has been updated to emit these enum tags as well eliding the now
   unneeded casts for each of the initialization arguments.

   Also added a Debug ASSERT to double check that we aren't consuming the union as one
   variant if it was initialized as another variant.

2. Factored FunctionPtr out of MacroAssemblerCodeRef.h into the WTF::TypedFunctionPtr class.
   TypedFunctionPtr is type safe with regards to the function prototype i.e. it preserved
   type info about the return value and arguments.

   TypedFunctionPtr can be used to call the function that it points to.

   We now use this for PropertySlot::GetValueFunc and PropertySlot::PutValueFunc for
   some added PAC diversity, and also as a debugging aid for the HashTableValue union
   variants.

3. Also introduced a WTF::FunctionPtr, which erases the function prototype info
   by always casting the stored pointer to a void(*)(int) pointer.

   FunctionPtr cannot be used to call the function that it points to (because its type
   Info is erased).

   In order to use it in a call, the FunctionPtr needs to be converted back into a
   normal function pointer (os a TypedFunctionPtr) first.

4. NativeFunction is now implemented using TypedFunctionPtr.

   RawNativeFunction has been removed.  Its role is subsumed by NativeFunction::Ptr.
   The only exception to this is one place in the JSC public API where it defines its own
   RawNativeFunction in its private implementation.  This was not changed to just so we
   don't have to #include wtf/FunctionPtr.h in that public API file.

   TaggedNativeFunction is now implemented using the type erased FunctionPtr.

5. CFunctionPtr (in MacroAssemblerCodeRef.h) has been removed.  It was only used by
   DOMJIT::Signature.  Instead, DOMJIT::Signature now defines its own FunctionPtr
   using WTF::FunctionPtr.

6. Removed all the aliases of GetValueFunc and PutValueFunc in PropertySlot, PutPropertySlot,
   etc.  instead, just have all the places use GetValueFunc and PutValueFunc directly.
   There's no value added by having all these aliases.

* Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
(JSC::JSCallbackObject<JSAPIWrapperObject>::getCallFunction):
(JSC::JSCallbackObject<JSAPIWrapperObject>::getConstructFunction):
* Source/JavaScriptCore/API/JSCallbackObject.cpp:
(JSC::JSCallbackObject<JSNonFinalObject>::getCallFunction):
(JSC::JSCallbackObject<JSNonFinalObject>::getConstructFunction):
(JSC::JSCallbackObject<JSGlobalObject>::getCallFunction):
(JSC::JSCallbackObject<JSGlobalObject>::getConstructFunction):
* Source/JavaScriptCore/API/glib/JSAPIWrapperGlobalObject.cpp:
(JSC::JSCallbackObject<JSAPIWrapperGlobalObject>::getCallFunction):
(JSC::JSCallbackObject<JSAPIWrapperGlobalObject>::getConstructFunction):
* Source/JavaScriptCore/API/glib/JSAPIWrapperObjectGLib.cpp:
(JSC::JSCallbackObject<JSAPIWrapperObject>::getCallFunction):
(JSC::JSCallbackObject<JSAPIWrapperObject>::getConstructFunction):
* Source/JavaScriptCore/assembler/LinkBuffer.h:
(JSC::LinkBuffer::link):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::readCallTarget):
(JSC::MacroAssemblerARM64::linkCall):
* Source/JavaScriptCore/assembler/MacroAssemblerCodeRef.h:
(JSC::toFunctionPtr):
(JSC::CFunctionPtr::CFunctionPtr): Deleted.
(JSC::CFunctionPtr::get const): Deleted.
(JSC::CFunctionPtr::address const): Deleted.
(JSC::CFunctionPtr::operator bool const): Deleted.
(JSC::CFunctionPtr::operator! const): Deleted.
(JSC::CFunctionPtr::operator== const): Deleted.
(JSC::CFunctionPtr::operator!= const): Deleted.
(JSC::FunctionPtr::FunctionPtr): Deleted.
(JSC::FunctionPtr::retagged const): Deleted.
(JSC::FunctionPtr::executableAddress const): Deleted.
(JSC::FunctionPtr::retaggedExecutableAddress const): Deleted.
(JSC::FunctionPtr::operator bool const): Deleted.
(JSC::FunctionPtr::operator! const): Deleted.
(JSC::FunctionPtr::operator== const): Deleted.
(JSC::FunctionPtr::operator!= const): Deleted.
(JSC::FunctionPtr<tag>::FunctionPtr): Deleted.
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::ftlThunkAwareRepatchCall):
(JSC::linkPolymorphicCall):
* Source/JavaScriptCore/create_hash_table:
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::compileFunction):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/domjit/DOMJITGetterSetter.h:
* Source/JavaScriptCore/domjit/DOMJITSignature.h:
(JSC::DOMJIT::Signature::Signature):
(JSC::DOMJIT::Signature::functionWithoutTypeCheck const):
* Source/JavaScriptCore/ftl/FTLLink.cpp:
(JSC::FTL::link):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::emitNakedNearCall):
(JSC::JIT::emitNakedNearTailCall):
* Source/JavaScriptCore/runtime/ExecutableBase.cpp:
(JSC::ExecutableBase::dump const):
* Source/JavaScriptCore/runtime/InternalFunction.cpp:
(JSC::InternalFunction::InternalFunction):
* Source/JavaScriptCore/runtime/JSCPtrTag.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.h:
* Source/JavaScriptCore/runtime/Lookup.h:
(JSC::HashTableValue::builtinGenerator const):
(JSC::HashTableValue::function const):
(JSC::HashTableValue::functionLength const):
(JSC::HashTableValue::propertyGetter const):
(JSC::HashTableValue::propertyPutter const):
(JSC::HashTableValue::domJIT const):
(JSC::HashTableValue::signature const):
(JSC::HashTableValue::accessorGetter const):
(JSC::HashTableValue::accessorSetter const):
(JSC::HashTableValue::constantInteger const):
(JSC::HashTableValue::lexerValue const):
(JSC::HashTableValue::lazyCellPropertyOffset const):
(JSC::HashTableValue::lazyClassStructureOffset const):
(JSC::HashTableValue::lazyPropertyCallback const):
(JSC::HashTableValue::builtinAccessorGetterGenerator const):
(JSC::HashTableValue::builtinAccessorSetterGenerator const):
* Source/JavaScriptCore/runtime/NativeFunction.h:
(JSC::toTagged):
(WTF::add):
(JSC::NativeFunction::NativeFunction): Deleted.
(JSC::NativeFunction::operator intptr_t const): Deleted.
(JSC::NativeFunction::operator bool const): Deleted.
(JSC::NativeFunction::operator! const): Deleted.
(JSC::NativeFunction::operator== const): Deleted.
(JSC::NativeFunction::operator!= const): Deleted.
(JSC::NativeFunction::operator()): Deleted.
(JSC::NativeFunction::rawPointer const): Deleted.
(JSC::TaggedNativeFunction::TaggedNativeFunction): Deleted.
(JSC::TaggedNativeFunction::operator bool const): Deleted.
(JSC::TaggedNativeFunction::operator! const): Deleted.
(JSC::TaggedNativeFunction::operator== const): Deleted.
(JSC::TaggedNativeFunction::operator!= const): Deleted.
(JSC::TaggedNativeFunction::operator()): Deleted.
(JSC::TaggedNativeFunction::operator NativeFunction): Deleted.
(JSC::TaggedNativeFunction::rawPointer const): Deleted.
* Source/JavaScriptCore/runtime/PropertySlot.h:
(JSC::PropertySlot::setCacheableCustom):
* Source/JavaScriptCore/runtime/PutPropertySlot.h:
(JSC::PutPropertySlot::setCustomValue):
(JSC::PutPropertySlot::setCustomAccessor):
(JSC::PutPropertySlot::customSetter const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::getHostFunction):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FunctionPtr.h: Added.
(WTF::TypedFunctionPtr::TypedFunctionPtr):
(WTF::TypedFunctionPtr::operator() const):
(WTF::TypedFunctionPtr::get const):
(WTF::TypedFunctionPtr:: const):
(WTF::TypedFunctionPtr::executableAddress const):
(WTF::TypedFunctionPtr::retaggedExecutableAddress const):
(WTF::TypedFunctionPtr::untaggedExecutableAddress const):
(WTF::TypedFunctionPtr::operator bool const):
(WTF::TypedFunctionPtr::operator! const):
(WTF::TypedFunctionPtr::operator== const):
(WTF::TypedFunctionPtr::operator!= const):
(WTF::TypedFunctionPtr::operator=):
(WTF::TypedFunctionPtr::encode):
(WTF::TypedFunctionPtr::decode):
(WTF::FunctionPtr::FunctionPtr):
(WTF::FunctionPtr::retagged const):
(WTF::add):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::createCrossOriginFunction):
(WebCore::JSDOMGlobalObject::createCrossOriginGetterSetter):
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::jsDOMWindowGetOwnPropertySlotRestrictedAccess):
(WebCore::JSDOMWindow::put):
* Source/WebCore/bindings/js/JSLocationCustom.cpp:
(WebCore::JSLocation::put):
* Source/WebCore/bindings/js/JSRemoteDOMWindowCustom.cpp:
(WebCore::JSRemoteDOMWindow::put):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHashTableValueArray):
* Source/WebCore/bindings/scripts/test/JS/JSDOMWindow.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSDedicatedWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSExposedStar.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSExposedToWorkerAndWindow.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSPaintWorkletGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSServiceWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSShadowRealmGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSSharedWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCEReactions.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCEReactionsStringifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestCallTracer.cpp:
(WebCore::JSTestCallTracerDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestClassWithJSBuiltinConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestConditionalIncludes.cpp:
(WebCore::JSTestConditionalIncludesDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditionallyReadWrite.cpp:
(WebCore::JSTestConditionallyReadWriteDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestDOMJIT.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSON.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONFilteredByExposed.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONIndirectInheritance.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInherit.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInheritFinal.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDelegateToSharedSyntheticAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestDomainSecurity.cpp:
(WebCore::JSTestDomainSecurityDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestEnabledBySetting.cpp:
(WebCore::JSTestEnabledBySettingDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestEnabledForContext.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestEventConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestEventTarget.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGenerateAddOpaqueRoot.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGenerateIsReachable.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestInterface.cpp:
(WebCore::JSTestInterfaceDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSTestInterfaceLeadingUnderscore.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestIterable.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestJSBuiltinConstructor.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyFactoryFunction.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyNoInterfaceObject.cpp:
(WebCore::JSTestLegacyNoInterfaceObjectPrototype::finishCreation):
* Source/WebCore/bindings/scripts/test/JS/JSTestLegacyOverrideBuiltIns.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestMapLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestMapLikeWithOverriddenOperations.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedAndIndexedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedDeleterWithIndexedGetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterCallWith.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedGetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterNoIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterThrowingException.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIdentifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIndexedGetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithIndexedGetterAndSetter.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyOverrideBuiltIns.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyUnforgeableProperties.cpp:
(WebCore::JSTestNamedSetterWithLegacyUnforgeablePropertiesDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestNamedSetterWithLegacyUnforgeablePropertiesAndLegacyOverrideBuiltIns.cpp:
(WebCore::JSTestNamedSetterWithLegacyUnforgeablePropertiesAndLegacyOverrideBuiltInsDOMConstructor::prototypeForStructure):
* Source/WebCore/bindings/scripts/test/JS/JSTestNamespaceConst.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNamespaceObject.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSTestOperationConditional.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestOverloadedConstructors.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestOverloadedConstructorsWithSequence.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestPluginInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestPromiseRejectionEvent.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReadOnlyMapLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReadOnlySetLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestReportExtraMemoryCost.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSerializedScriptValueInterface.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSetLike.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestSetLikeWithOverriddenOperations.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifier.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierAnonymousOperation.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierNamedOperation.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierOperationImplementedAs.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierOperationNamedToString.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierReadOnlyAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestStringifierReadWriteAttribute.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestTaggedWrapper.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSTestTypedefs.cpp:
(WebCore::JSTestTypedefsDOMConstructor::construct):
* Source/WebCore/bindings/scripts/test/JS/JSWorkerGlobalScope.cpp:
* Source/WebCore/bindings/scripts/test/JS/JSWorkletGlobalScope.cpp:
* Source/WebCore/cssjit/FunctionCall.h:
(WebCore::FunctionCall::FunctionCall):
(WebCore::FunctionCall::setFunctionAddress):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementAttributeFunctionCallValueMatching):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementFunctionCallTest):

Canonical link: https://commits.webkit.org/253414@main
